### ISSUE TITLE
Using crash in a lambda crashes the app without the function being called

### ISSUE THREAD

**Issue by cscalfani** [purescript/purescript#4379]
## To Reproduce

Steps to reproduce the behavior.

## PureScript version

0.x.x

**Comment by cscalfani** [comment]
For anyone who is running into this issue, replacing:

```
let x = \_ -> unsafePartial $ crash "this is an error"
```

with:

```
let x = \_ -> unsafeCrashWith "this is an error"
```

fixes the problem (thanks to @natefaubion for this suggestion) and is a great workaround until this is fixed

**Comment by rhendric** [comment]
There's definitely a bug here (seems likely to be an unforeseen consequence of #3915), but based on the fact that you're providing a string message you probably want `crashWith` (`crash` is fully polymorphic and is happy to pretend to be a function that accepts your string argument, but nothing will happen to it), which doesn't suffer from this problem either.

**Comment by rhendric** [comment]
I don't know what the best resolution to this would be. #3915 was founded on the assumption that eliminating a class constraint by supplying a class dictionary is a pure operation, and `crash` is a counterexample. I would hate to lose #3915 over this, but it's not clear to me that it'd be a good idea to, say, special-case `Partial` constraints as impure in this way. Why should `Partial` get special treatment over any other class a user might define and expect the same properties from?

My gut feeling is that the design of `crash` is a mistake in a call-by-value language, but that's easy for me to say as someone who doesn't use it—and indeed, someone who feels that `Partial` is generally more trouble than it's worth. Can anyone make an argument for when class constraints should and should not be considered pure, considering that if we can't ever consider them to be pure then #3915 needs to be reverted for good?

**Comment by garyb** [comment]
Would the fact that it's a nullary class be a special enough case for? Similar to the fact `Partial` is weird, nullary type classes are generally.

**Comment by natefaubion** [comment]
Is the issue about whether the scope hoisting can cause premature evaluation of user code? For example, one reason we don't want to hoist arbitrary evaluation up in a strict setting is that code can be intentionally deferred in such a way that correctness depends on it, otherwise you get non-termination. Is it possible to construct a program (other than `Partial`) that is correct under the previous type-class insertion, which is made incorrect in the hoisted version? If so, is it that we need to make sure that the transformation still preserves the deferred property of the original code?

**Comment by natefaubion** [comment]
Another alternative is just hoisting the allocation of dictionaries, but not the application of user code itself. I suspect this is the larger source of improvement, but I obviously don't have benchmarks.

**Comment by natefaubion** [comment]
In addition to that, we could hoist applications to user code only to the closest, shared, dominating scope, rather than as high as possible. I think that would preserve existing expectations on deferred-ness. In the `crash` example, it would not result in the application to `crash` being lifted outside of the lambda it's in because there is no other application of that function in a scope that dominates the lambda scope.

**Comment by rhendric** [comment]
> Would the fact that it's a nullary class be a special enough case for? Similar to the fact `Partial` is weird, nullary type classes are generally.

If `Partial` had an argument we would still have the same problem. We could arbitrarily say that nullary classes have their own semantics and that would be a way forward, but I find it a little undermotivated as a solution. What are the other ways that nullary type classes are weird?

> Is it possible to construct a program (other than `Partial`) that is correct under the previous type-class insertion, which is made incorrect in the hoisted version?

Doubt I could provide a mathematical proof, but if you exclude any as-named unsafe constructs then I don't think so. Assuming the type system is doing its job, the only incorrect behavior we're talking about is non-termination. There's a specific carve-out in the CSE code that prevents recursive bindings from being floated out of their scopes, so the question becomes whether there's a way for a non-recursive binding to have different termination semantics if it's floated to the outermost scope in which it could be defined. I can't think of any way that could happen.

The issue with `crash` is that it causes a premature side effect not captured by the type system if the `Partial` constraint doesn't act as a defer. Mere non-termination via an infinite loop wouldn't be enough. The question is whether code can make the assumption that a constraint effectively grants a binding lazy semantics, such that its definition is guaranteed not to be executed until a use of the binding is dereferenced.

> Another alternative is just hoisting the allocation of dictionaries, but not the application of user code itself.

That's an option, sure. Even this is an observable change with respect to `Partial` code, though; contrived example follows:

```purs
class Foo a where
  poison :: a

instance Partial => Foo Int where
  poison = crashWith "don't use this"

instance Foo String where
  poison = "poison"

bar :: forall a. Foo a => a -> Boolean
bar _ = true

main :: Effect Unit
main = do
  let x = \_ -> unsafePartial $ bar 1
  log "done"
```

Under the historical semantics, this would terminate normally, because the `Foo Int` dictionary wouldn't be constructed unless `x` is invoked. Floating dictionary allocations is enough to cause the crash to happen.

(Granted, use of this pattern is probably somewhere between ‘exactly never’ and an order of magnitude more rare than garden-variety use of `crash`. But what other use cases might I not be thinking of?)

> In addition to that, we could hoist applications to user code only to the highest, shared, dominating scope, rather than as high as possible.

I agree that would be safe, but my suspicion is that this is too conservative to do much for performance, either on its own or on top of the other proposal.

**Comment by cscalfani** [comment]
> Is it possible to construct a program (other than Partial) that is correct under the previous type-class insertion, which is made incorrect in the hoisted version?

Yes it is. Stack overflows are a non-Partial example of the problems of generating overly eager code:

```haskell
fact ∷ Int → Int
fact x = x * fact x - 1

class Foo a where
  poison :: a

instance Foo Int where
  poison = fact 100000 -- will overflow stack if executed

instance Foo String where
  poison = "poison"

bar :: ∀ a. Foo a => a -> Boolean
bar _ = true

main :: Effect Unit
main = do
  let x = \_ -> bar 1
  log "done"
```

Even if this didn't overflow the stack, the computationally expensive, premature or, in this case, unnecessary call to `fact` can have a major performance impact.

**Comment by rhendric** [comment]
That example blows the stack even in old versions of PureScript. The `Foo Int` instance, having no constraints, was always constructed eagerly at the top level. If you add a constraint of any sort to the instance, then you have a valid example. But note that the constraint doesn't have to be `Partial` or even a nullary class—any constraint would do.

So the question again: should code be able to make the assumption that adding a constraint necessarily means fully lazifying a binding, guaranteeing that its definition isn't evaluated until a use of the binding is dereferenced?

**Comment by cscalfani** [comment]
Confirming @rhendric's comment:
>  If you add a constraint of any sort to the instance, then you have a valid example

The following change:
```haskell
instance Show a => Foo Int where
  poison = fact 100000
```
produces:
```javascript
var fooInt = function (dictShow) {
    return {
        poison: fact(100000)
    };
};
```

**Comment by garyb** [comment]
> What are the other ways that nullary type classes are weird?

By "weird" I guess I just meant extremely uncommon - I thought `Partial` and `CallByName` were the only ones I knew of, but turns out `CallByName` isn't nullary either :wink:.

**Comment by natefaubion** [comment]
> In 0.14.x this would not crash, but in 0.15.x the dictionary passed in the hoisted call will crash.

A total program that blows the stack on a particular runtime is different from a non-terminating program. I understand that _when_ that crashes is different, but it would crash regardless on a JS runtime. It may very well not crash at all on a different runtime depending on the definition of `fact` or the particular constraints of the runtime (just bump the stack limit!). I don't really consider that an example of correctness _per se_, as it depends on something that's necessarily undefined behavior for our purposes. Additionally, this kind of example fails to provide a justification for the previous behavior, because presumably if dictionary instantiation was really expensive, you _definitely_ would want it to be hoisted and shared.

> So the question again: should code be able to make the assumption that adding a constraint necessarily means fully lazifying a binding, guaranteeing that its definition isn't evaluated until a use of the binding is dereferenced?

I think this is hard to justify. Given that constraint abstraction, resolution (ie, which dictionary is actually used), and application is completely implicit and outside of my control as a user, I find it difficult to say I should be able to reason about it in my own code. I think if you have explicit demands around evaluation, then the type should reflect that, and I don't believe constraints _in general_ serve that purpose. And I say that as the author of CallByName (which _does_ carry an explicit warning!). I think it's fair to say that users can reason about evaluation of user-written applications under lambdas, and that we would never prematurely evaluate a user written application.

I think that:
```purescript
crash :: forall a. Partial => a
crash = crashWith "Partial.crash: partial function"
```

Is a problematic definition, because there are no lambda abstractions, and we should just remove it.

```purescript
crashWith :: forall a. Partial => String -> a
```
Is totally fine because it uses an explicit abtraction.

All that is to say, I think a "won't fix" for this is justifiable, and that `crash` on it's own is just a problematic API (not just for this reason).

**Comment by cscalfani** [comment]
I think a "won't fix" is fair as well as the removal of `crash`, but I'd also advocate a warning for future developers to not write functions in the following form:

```haskell
myCrash :: forall a. Partial => a
```

With an explanation as to why or at least a link to this issue.

I think it would be helpful to have that documentation at least here: https://pursuit.purescript.org/builtins/docs/Prim#t:Partial
and possibly even here: https://pursuit.purescript.org/packages/purescript-partial/4.0.0

**Comment by MonoidMusician** [comment]
Personally I'd like to keep the semantics of `Partial => _` as being `Unit -> _` if possible.

Can't we just fix this by removing the [inlining of `unsafePartial`](https://github.com/rhendric/purescript/blob/283c05766448e07bb8bee721d217c5fe78478670/src/Language/PureScript/CoreImp/Optimizer/Inliner.hs#L316)?

There should never be a `Partial` constraint to apply at the top-level, so lifting should not occur at the top-level. If you're inside a lambda with a `Partial`, it should be safe to lift within that scope, right?

**Comment by natefaubion** [comment]
> Can't we just fix this by removing the [inlining of unsafePartial](https://github.com/rhendric/purescript/blob/283c05766448e07bb8bee721d217c5fe78478670/src/Language/PureScript/CoreImp/Optimizer/Inliner.hs#L316)?

Removing the inlining of `unsafePartial` is problematic, as it breaks other compiler optimizations (such as unsafe array indexing and tco).

I don't think the fact that it's Partial is what matters. What if I implement my own version of Partial that's more descriptive?

**Comment by rhendric** [comment]
If you mean something like this:
```purs
main = unsafePartial do
  let x = \_ -> crash
  pure unit
```
then no, this is still unsafe as long as `crash` can leave the definition of `x`, even if it's forbidden from escaping the argument to `unsafePartial`.

**Comment by MonoidMusician** [comment]
Hmm okay. I don't know then.

Removing the inlining doesn't actually do anything. I guess nullary dictionaries are completely disregarded in CoreFn. Certainly breaks my mental model of how `Partial` works, I would have expected that at least some variable reference was maintained for scoping even if its value was `undefined`. Which means that user-made nullary typeclasses would have the same issues.

**Comment by natefaubion** [comment]
> Certainly breaks my mental model of how Partial works, I would have expected that at least some variable reference was maintained for scoping even if its value was undefined. Which means that user-made nullary typeclasses would have the same issues.

`Partial` is an unusable constraint, it's not possible to make anything in particular demand a specific instance in scope. You would need to go the reflection route and use a rank-2 polymorphic type and `Proxy` to accomplish this. That's part of the reason why it's so dodgy (not unlike  `Given`).

**Comment by natefaubion** [comment]
Just to spitball, I think potentially more well-behaved version of `Partial` would be to make it an actual Nominal `Type` like `Effect`, where a partial case elaborates to some `Partial a`.

```purescript
example1 :: Maybe Int -> Int
example1 = case _ of
  Just a -> a
  Nothing -> 0

example2 :: Maybe Int -> Partial Int
example2 = case _ of
  Just a -> a
```

That is, the case body (everything after `of`) elaborates to an unsafe closure. Then `unsafePartial` is equivalent to `unsafePerformEffect`, and you can do all the usual compositional things you can do with actual types, vs constraints. This also solves issues I've had where it's really easy to have a `Partial` constraint that's _too_ broad.

```purescript
example :: Partial => Int
example = do
  let
    a = case ... of
      Just x -> x
    b = case ... of
      Just y -> y
      Nothing -> 0
  a + b
```
If I only want `a` to be partial, I'm out of luck. I could easily miss the `Nothing` case in `b` and I wouldn't be told about it. I would need to non-trivially refactor my code.

**Comment by rhendric** [comment]
Yeah, that's basically a restricted `Effect`—maybe the sort of thing that the old `Eff` type was good at representing?

I think the (tempting, dangerous) appeal of `Partial` over that encoding is how the whole monad machinery doesn't need to be brought in to handle result values. If `unsafePartial` did actually delimit the partiality of the computation, it'd be a slam dunk; the fact that it does not is the reason I try to avoid it altogether.

We can probably file this safely under ‘things we'll never do’, but my dream solution might be something like a `Prim.PartialFunction :: Type -> Type -> Type` type (`infixr 0 type PartialFunction as ->?`, perhaps) with special language integration:
* A lambda term can be typed as a `Function` or as a `PartialFunction`, defaulting to the former if the following rules allow it.
* `e1 :: PartialFunction a b`, `e2 :: a` ⊢ `e1 e2 :: b` but requires the nearest enclosing lambda abstraction to be typed as a `PartialFunction` (and is illegal outside of any abstraction).
* A `case` expression with missing clauses is likewise only legal directly inside a `PartialFunction`.
* A `Function` can be implicitly coerced to a `PartialFunction` with the same type arguments (but not, of course, vice versa).
* Challenge: do-notation. (I might imagine a parallel `FunctorP` up to `MonadP` class hierarchy that offer `mapP :: (a ->? b) -> f a ->? f b`, `bindP :: (a ->? f b) -> f a ->? f b`, etc. variants, and using them as appropriate when partial application is present in a do block. Note that not every `Functor` is a `FunctorP`, etc. Someone cleverer than I might come up with something more elegant.)

This wouldn't add any run-time thunks and should be transparent to the user, unlike a monadic interface. An `unsafeApplyPartial :: PartialFunction a b -> a -> b` would be able to offer actual guarantees about when the partial part of the program is complete (to wit: after it has been applied to two arguments), unlike the existing `Partial` class. But it'd probably be a headache getting this to be compatible with the existing type system, based on the folklore I'm aware of about mixing implicit coercion/subtypes with HM.

**Comment by jy14898** [comment]
> We can probably file this safely under ‘things we'll never do’, but my dream solution might be something like a Prim.PartialFunction :: Type -> Type -> Type type (infixr 0 type PartialFunction as ->?, perhaps) with special language integration:

You could do something like generalize `Function` with a totality type tag, which lets you then track totality and write totality polymorphic combinators. Then, we get to use all our normal functions like bind as we normally do. I think this is even more of a less likely direction than what you suggest, however

```purescript
foreign import data Totality :: Type
foreign import data Total :: Totality
foreign import data Partial :: Totality

foreign import data FUNCTION :: Totality -> Type -> Type -> Type

type Function = FUNCTION Total

-- for a lack of a good mixfix syntax...
applyN :: forall ... (t :: Totality). (FUNCTION t a a) -> Int -> FUNCTION t a a

-- Compiler now requires crash to be behind some abstraction
crash :: FUNCTION Partial Unit a

-- Error: partial function application must be inside a partial abstraction
f = let _ = crash unit in 0

--  Compiler infers f :: FUNCTION Partial a (b -> Int)
f _ = let _ = crash unit in \_ -> 0
```
